collector/terminal_services.go

```diff
@@ -0,0 +1,403 @@
+)
+
+const ConnectionBrokerFeatureID uint32 = 133
+
+func init() {
+	registerCollector("terminal_services", NewTerminalServicesCollector, "Terminal Services", "Terminal Services Session", "Remote Desktop Connection Broker Counterset")
+}
+
+var (
+	connectionBrokerEnabled = isConnectionBrokerServer()
+)
+
+type Win32_ServerFeature struct {
+	ID uint32
+}
+
+func isConnectionBrokerServer() bool {
+	var dst []Win32_ServerFeature
+	q := queryAll(&dst)
+	if err := wmi.Query(q, &dst); err != nil {
+		return false
+	}
+	for _, d := range dst {
+		if d.ID == ConnectionBrokerFeatureID {
+			return true
+		}
+	}
+	log.Debug("host is not a connection broker skipping Connection Broker performance metrics.")
+	return false
+}
+
+// A TerminalServicesCollector is a Prometheus collector for WMI
+// Win32_PerfRawData_LocalSessionManager_TerminalServices &  Win32_PerfRawData_TermService_TerminalServicesSession  metrics
+// https://docs.microsoft.com/en-us/previous-versions/aa394344(v%3Dvs.85)
+// https://wutils.com/wmi/root/cimv2/win32_perfrawdata_localsessionmanager_terminalservices/
+type TerminalServicesCollector struct {
+	LocalSessionCount           *prometheus.Desc
+	ConnectionBrokerPerformance *prometheus.Desc
+	HandleCount                 *prometheus.Desc
+	PageFaultsPersec            *prometheus.Desc
+	PageFileBytes               *prometheus.Desc
+	PageFileBytesPeak           *prometheus.Desc
+	PercentPrivilegedTime       *prometheus.Desc
+	PercentProcessorTime        *prometheus.Desc
+	PercentUserTime             *prometheus.Desc
+	PoolNonpagedBytes           *prometheus.Desc
+	PoolPagedBytes              *prometheus.Desc
+	PrivateBytes                *prometheus.Desc
+	ThreadCount                 *prometheus.Desc
+	VirtualBytes                *prometheus.Desc
+	VirtualBytesPeak            *prometheus.Desc
+	WorkingSet                  *prometheus.Desc
+	WorkingSetPeak              *prometheus.Desc
+}
+
+// NewTerminalServicesCollector ...
+func NewTerminalServicesCollector() (Collector, error) {
+	const subsystem = "terminal_services"
+	return &TerminalServicesCollector{
+		LocalSessionCount: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "local_session_count"),
+			"Number of Terminal Services sessions",
+			[]string{"session"},
+			nil,
+		),
+		ConnectionBrokerPerformance: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "connection_broker_performance_total"),
+			"The total number of connections handled by the Connection Brokers since the service started.",
+			[]string{"connection"},
+			nil,
+		),
+		HandleCount: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "handle_count"),
+			"Total number of handles currently opened by this process. This number is the sum of the handles currently opened by each thread in this process.",
+			[]string{"session_name"},
+			nil,
+		),
+		PageFaultsPersec: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "page_fault_per_sec"),
```

Comment:
Is this actually per_sec or is the raw data a counter?

Reply:
Yes it looks like total and counter. i will change it.